test/tls13encryptiontest.c

```diff
@@ -0,0 +1,304 @@
+{
+    int ret = 0;
+    unsigned char *refd;
+    long refdatalenl;
+    size_t refdatalen;
+
+    if (enc)
+        refd = OPENSSL_hexstr2buf(refdata[recnum].ciphertext, &refdatalenl);
+    else
+        refd = OPENSSL_hexstr2buf(refdata[recnum].plaintext, &refdatalenl);
+
+    if (refd == NULL) {
+        fprintf(stderr, "Failed to get reference data for record %"OSSLzu
+                         "\n", recnum);
+        goto err;
+    }
+    refdatalen = (size_t)refdatalenl;
+
+    if (rec->length != refdatalen) {
+        fprintf(stderr, "Unexpected length for record %"OSSLzu"\n", recnum);
+        goto err;
+    }
+
+    if (memcmp(rec->data, refd, refdatalen) != 0) {
+        fprintf(stderr, "Data does not match for record %"OSSLzu"\n", recnum);
+        goto err;
+    }
+
+    ret = 1;
+
+ err:
+    OPENSSL_free(refd);
+    return ret;
+}
+static int test_tls13_encryption(void)
+{
+    SSL_CTX *ctx = NULL;
+    SSL *s = NULL;
+    SSL3_RECORD rec;
+    unsigned char *key = NULL, *iv = NULL, *seq = NULL;
+    const EVP_CIPHER *ciph = EVP_aes_128_gcm();
+    int ret = 0;
+    size_t ivlen, ctr;
+
+    rec.data = NULL;
+
+    ctx = SSL_CTX_new(TLS_method());
+    if (ctx == NULL) {
+        fprintf(stderr, "Failed creating SSL_CTX\n");
+        goto err;
+    }
+
+    s = SSL_new(ctx);
+    if (s == NULL) {
+        fprintf(stderr, "Failed creating SSL\n");
+        goto err;
+    }
+
+    s->enc_read_ctx = EVP_CIPHER_CTX_new();
+    s->enc_write_ctx = EVP_CIPHER_CTX_new();
+    if (s->enc_read_ctx == NULL || s->enc_write_ctx == NULL) {
+        fprintf(stderr, "Failed creating EVP_CIPHER_CTX\n");
+        goto err;
+    }
+
+    for (ctr = 0; ctr < OSSL_NELEM(refdata); ctr++) {
+        /*
+         * Load the record, set up the read/write sequences and load the key into
+         * the EVP_CIPHER_CTXs
+         */
+        ivlen = EVP_CIPHER_iv_length(ciph);
+        if (!load_record(&rec, ctr, &key, s->read_iv, ivlen,
+                         RECORD_LAYER_get_read_sequence(&s->rlayer))) {
+            fprintf(stderr, "Failed loading key into EVP_CIPHER_CTX\n");
+            goto err;
+         }
+
+        memcpy(RECORD_LAYER_get_write_sequence(&s->rlayer),
+               RECORD_LAYER_get_read_sequence(&s->rlayer), SEQ_NUM_SIZE);
+        memcpy(s->write_iv, s->read_iv, ivlen);
+
+        if (EVP_CipherInit_ex(s->enc_write_ctx, ciph, NULL, key, NULL, 1) <= 0
+                || EVP_CipherInit_ex(s->enc_read_ctx, ciph, NULL, key, NULL, 0)
+                   <= 0) {
+            fprintf(stderr, "Failed loading key into EVP_CIPHER_CTX\n");
+            goto err;
+        }
+
+        /* Encrypt it */
+        if (tls13_enc(s, &rec, 1, 1) != 1) {
+            fprintf(stderr, "Failed to encrypt record\n");
+            goto err;
+        }
+
+        if (!test_record(&rec, ctr, 1)) {
+            fprintf(stderr, "Record encryption test failed\n");
+            goto err;
+        }
+
```

Comment:
I think you mean the one between tls13_enc() and test_record(), rather than this one...at least that makes more sense to me in terms of paragraphs of code - so I removed that one.